Fix NRE in CA1825: AvoidZeroLengthArrayAllocations

jamesqo opened the pull request:
Fixes #1298 

The original case was failing because it assumed the type of the declaration would match up with the type of the expression. e.g. for `object[] a = new string[0]`, the codefix would be `Array.Empty<object>()`. This PR changes that approach so it doesn't look at the declaration, and determines the element type solely from the expression using IOperation APIs.

/cc @333fred

jamesqo commented on src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/AvoidZeroLengthArrayAllocationsTests.cs:
```diff
@@ -273,29 +280,13 @@ End Class
     class C
     {
-	    public static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+        public static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
```
Changes tab to spaces

mavasani commented on src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.Fixer.cs:
```diff
@@ -46,20 +48,50 @@ public sealed override FixAllProvider GetFixAllProvider()
 
-            SyntaxNode variableDecl = generator.GetDeclaration(nodeToFix, DeclarationKind.Variable);
-            SyntaxNode typeNode = generator.GetType(variableDecl);
+            var operation = semanticModel.GetOperationInternal(nodeToFix, cancellationToken);
```
Why is this calling `GetOperationInternal` instead of `GetOperation`?

jamesqo replied:
I saw some other code that used the `Internal` overloads of the IOperation APIs, so I followed. e.g. There are a lot of `RegisterOperationActionInternal` calls.

mavasani replied:
Interesting, I thought I recently removed all the Internal APIs and calls. Did you sync your sources with latest master?

mavasani replied:
It should have been deleted with https://github.com/dotnet/roslyn-analyzers/pull/1334/files#diff-5cbfbc6801faa8ac7bc5ae144689bb70L155. Please resync to latest.

mavasani replied:
A simpler way to implement this would be to use [FindNode](http://source.roslyn.io/#Microsoft.CodeAnalysis/Syntax/SyntaxNode.cs,1973a2affa06beb8,references) API for the diagnostic's location span, and then request the type associated with the array creation using `SemanticModel.GetTypeInfo(node)` and check if it is an array type. This will avoid you the trouble of having to figure out what operation node maps to the diagnostic span.

jamesqo commented on src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.Fixer.cs:
```diff
@@ -46,20 +48,50 @@ public sealed override FixAllProvider GetFixAllProvider()
+                    // Happens when the array is converted to a different type as it's created.
+                    // e.g. M's signature is M(object obj), M(new object[0])
+                    //      In this case, the IConversionExpression is the child of the IArgument.
+                    return FindArrayCreation(
```
@333fred Referring to the email you sent, do you mean the changes to `GetOperation` would make this unnecessary? If so should I add a TODO to revisit this?

333fred replied:
It should still be necessary in the case of an explicit cast as an argument to a function, for example `M((object)new object[10])`. Could you add a test case for that, and also a test case for something like `M((string[])new object[0])`, to make sure we're supplying the correct generic argument.

---
In reply to: [149532885](https://github.com/dotnet/roslyn-analyzers/pull/1339#discussion_r149532885) [](ancestors = 149532885)

333fred commented:
@jamesqo looks like the build failures are because you haven't merged master recently.

333fred commented on src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.Fixer.cs:
```diff
@@ -46,20 +48,50 @@ public sealed override FixAllProvider GetFixAllProvider()
-            SyntaxNode variableDecl = generator.GetDeclaration(nodeToFix, DeclarationKind.Variable);
-            SyntaxNode typeNode = generator.GetType(variableDecl);
+            var operation = semanticModel.GetOperationInternal(nodeToFix, cancellationToken);
+            var arrayCreationExpression = FindArrayCreation(operation);
```
Please use explicit type when the type is not apparent from the right-hand side of an expression.

333fred commented on src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.Fixer.cs:
```diff
@@ -46,20 +48,50 @@ public sealed override FixAllProvider GetFixAllProvider()
+                        ((IConversionExpression)operation).Operand);
+            }
+
+            return null;
```
This should never get hit, as that would mean we flagged something incorrectly. Please add an assert for this.

jamesqo replied:
@333fred I noticed in the Roslyn repo, they use `throw ExceptionUtilities.Unreachable` when something shouldn't get hit. Is there something like that in this repo?

333fred replied:
We don't currently, but you could port it over from the roslyn repo over to Analyzer.Utilities.

jamesqo commented:
> Lets avoid operations in the fixer code to keep it simpler.

That would require me to make different C#/VB classes for the fixer in order to access the language-specific syntax nodes. IMO, that sounds more complex.

mavasani commented:
> That would require me to make different C#/VB classes for the fixer in order to access the language-specific syntax nodes. IMO, that sounds more complex.

Not really. Both the APIs I mentioned lie in the language agnostic layer: [FindNode](http://source.roslyn.io/#Microsoft.CodeAnalysis/Syntax/SyntaxNode.cs,1973a2affa06beb8) and [GetTypeInfo](http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/Extensions.cs,d90409d281b869e5,references)

jamesqo commented:
@mavasani Yes, I just saw your comment. I previously tried that, actually: see [here](https://github.com/dotnet/roslyn-analyzers/pull/1339/commits/2789f84d78bf9410a3e63a86ef70931d4b3d01c1). It didn't work for calls to methods like `M(new object[0])`, though. `FindNode` would return an `ArgumentSyntax`, and `GetTypeInfo` wouldn't work (in the code I linked to, `arrayType` would be null).

mavasani commented:
> It didn't work for calls to methods like M(new object[0]), though. FindNode would return an ArgumentSyntax

Did you try to pass `getInnermostNodeForTie = true` to the call to [FindNode](http://source.roslyn.io/#Microsoft.CodeAnalysis/Syntax/SyntaxNode.cs,808)?

mavasani commented:
@jamesqo Did the workaround help?

jamesqo commented:
@mavasani I was unexpectedly busy until this Wednesday because I was working on a paper. I'll apply your workaround sometime today and get back to you on this.

jamesqo commented:
@mavasani I just tried your workaround (see the latest commit). There seems to be a problem for cases like this:

```cs
byte[] arr = { };
```

The IOperation APIs manage to correctly infer the type of the `{ }`. However, `semanticModel.GetTypeInfo` can't infer it and the code nullrefs at `arrayType.ElementType`.

jamesqo commented:
@mavasani Can I switch back to the IOperation impl, or do you have another suggestion on how I can stick to the SyntaxNode-based approach?

mavasani commented:
James, this seems to be a bug in Roslyn for the GetTypeInfo API. Kindly file an issue on Roslyn and a null check for the code in the analyzer that uses GetTypeInfo.Type.

jamesqo commented:
@mavasani OK, I addressed your feedback and opened dotnet/roslyn#23545. Do people here do GitHub squash merges, or should I manually squash my commits?

mavasani commented:
@jamesqo Thank you for the contribution and trying the different approaches!

> Do people here do GitHub squash merges, or should I manually squash my commits?

We generally create a merge commit, I will do the same for your PR.

mavasani commented:
@333fred Can you please take a quick look at the changes too? If you are good, we can merge.

mavasani commented:
I am merging this to avoid dupes. @333fred Feel free to open a separate issue for any further changes.
